### PR TITLE
Gather_nd

### LINKED ISSUE #22160: gather_nd

### PR DESCRIPTION
**Pull request by Ambro19**
Closes #22160

Added gather_nd function into manipulation.py file. This function is actually a high-dimensional extension of gather function and supports for simultaneous indexing by multiple axes.

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by Ambro19** [comment]
Hey@vaithak, I am trying to get to test_paddle_manipultion.py to test the function. Any help? Thanks

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 1]
```diff
@@ -8,3 +8,34 @@
 @with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
 def layer_norm(x, normalized_shape, weight=None, bias=None, epsilon=1e-05, name=None):
     return ivy.layer_norm(x, normalized_shape, weight, bias, epsilon)
+
+
+# batch_norm
+@to_ivy_arrays_and_back
+@with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
+def batch_norm(
```
Comment (Ambro19) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:16]: Added this batch_norm function.

File: ivy/functional/frontends/paddle/tensor/manipulation.py [hunk 2]
```diff
@@ -146,5 +147,38 @@ def roll(x, shifts, axis=None, name=None):
...
 def take_along_axis(arr, indices, axis):
     return ivy.take_along_axis(arr, indices, axis)
+
+
+@with_supported_dtypes(
+    {"2.5.1 and below": ("bool", "float32", "float64", "int32", "int64")},
+    "paddle",
+)
+@to_ivy_arrays_and_back
+def gather_nd(x, index, name=None):
```
Comment (Ambro19) [inline ivy/functional/frontends/paddle/tensor/manipulation.py:163]: Added this gather_nd function.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py [hunk 3]
```diff
@@ -43,3 +43,38 @@ def test_paddle_layer_norm(
...
+
+# bach_norm
+@handle_frontend_test(
+    fn_tree="paddle.nn.functional.bach_norm",
+    values_tuple=_generate_data_layer_norm(
+        available_dtypes=helpers.get_dtypes("float"),
+    ),
+    eps=st.floats(min_value=0.01, max_value=0.1),
+)
+def test_paddle_bach_norm(
```
Comment (Ambro19) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:56]: Added its testing function.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py [hunk 4]
```diff
@@ -637,3 +637,40 @@ def test_paddle_take_along_axis(
...
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("numeric"),
+        min_num_dims=2,
+        min_dim_size=2,
+    ),
+    shift=helpers.ints(min_value=1, max_value=10),
+    axis=helpers.ints(min_value=-1, max_value=1),
+    test_with_out=st.just(False),
+)
+def test_paddle_gather_nd(
```
Comment (Ambro19) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:654]: Added its testing function.

**Review by Ambro19 — COMMENTED** [review]
There are 4 changes added. Please review when you can and let me if changes are needed. Thanks

**Review by vaithak — CHANGES_REQUESTED** [review]
Hi @Ambro19, your PR contains some changes which are already present on the main, can you please resolve the conflicts to make sure only new changes are present in the PR.

Also, for batch norm you ready have a PR open here: https://github.com/unifyai/ivy/pull/22157, which is already under review.   Let's not add those changes in this PR too.

**Comment by Ambro19** [comment]
Hi @vaithak, I fixed the conflicting merge and more as you required. Please take a look at and let me know. Also, sorry for the additional request. I wasn't sure. Thanks.

**Comment by Ambro19** [comment]
Hi @vaithak, I fixed the conflicting merge and more as you required. Please take a look at and let me know. Also, sorry for the additional request. I wasn't sure :) Thanks.

File: ivy/functional/frontends/paddle/nn/functional/norm.py (hunk 1 above)
Comment (vaithak) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:16]: As mentioned earlier, this can be removed as you already have a separate PR for this

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py (hunk 3 above)
Comment (vaithak) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:56]: same here

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py [hunk 5]
```diff
@@ -718,35 +718,12 @@ def test_paddle_take_along_axis(
...
-    *,
-    dtype_m_k_axes,
-    on_device,
-    fn_tree,
-    frontend,
-    backend_fw,
-    test_flags,
-):
-    input_dtype, m, k, axes = dtype_m_k_axes
+Solved from the merge of gather_nd and main branches.
```
Comment (vaithak) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:721]: these comments can be removed

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py [hunk 6]
```diff
@@ -718,35 +718,12 @@ def test_paddle_take_along_axis(
...
         backend_to_test=backend_fw,
         frontend=frontend,
         test_flags=test_flags,
         fn_tree=fn_tree,
         on_device=on_device,
-        x=m,
-        k=k,
-        axes=tuple(axes),
-    )
+Again, Solved from the merge of gather_nd and main branches.
```
Comment (vaithak) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:729]: same for this comment

**Review by vaithak — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py [hunk 7]
```diff
@@ -708,3 +713,30 @@ def test_unstack(
...
         keepdims=keepdims,
     )
+
+
+# gather_nd
+@handle_test(
+    fn_tree="paddle.gather_nd",
+    dtype_param_and_indices=_gather_helper(),
+)
+def test_paddle_gather_nd(
```
Comment (vaithak) [inline ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py:723]: the tests are failing on the ci, can you run this locally to ensure it passes? Thanks

Comment (Ambro19) [reply ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py]: Hi @vaithak, sorry I just see your comment now. I will work on it and let you know. Thanks.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py (hunk 6 above)
Comment (Ambro19) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py]: Ok, no problem. I will take a looks at it. Thank you.

File: ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py (hunk 7 above)
Comment (Ambro19) [reply ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py]: image

Update: Above is this output when running test on gather_nd. I am looking more into the error to fix it.

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 8]
```diff
@@ -10,9 +10,4 @@ def layer_norm(x, normalized_shape, weight=None, bias=None, epsilon=1e-05, name=
     return ivy.layer_norm(x, normalized_shape, weight, bias, epsilon)
 
 
-@to_ivy_arrays_and_back
-@with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
-def normalize(x, p=2, axis=1, epsilon=1e-12, name=None):
-    if axis < 0:
-        axis = ivy.get_num_dims(x) + axis
-    return ivy.lp_normalize(x, p=p, axis=axis)
+Removed conflicts.
```
Comment (vaithak) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:13]: Please don't remove any code snippet unrelated to your changes.

File: ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py [hunk 9]
```diff
@@ -775,19 +780,5 @@ def test_unstack(
...
-def test_zero_pad(*, dtype_value_pad_width, test_flags, backend_fw, fn_name, on_device):
-    dtype, value, pad_width = dtype_value_pad_width
-    helpers.test_function(
-        input_dtypes=dtype,
-        test_flags=test_flags,
-        backend_to_test=backend_fw,
-        fn_name=fn_name,
-        on_device=on_device,
-        x=value[0],
-        pad_width=pad_width,
```
Comment (vaithak) [inline ivy_tests/test_ivy/test_functional/test_core/test_manipulation.py:792]: same for this as well

File: ivy/functional/frontends/paddle/nn/functional/norm.py (hunk 8 above)
Comment (Ambro19) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:13]: Hi, Oh I so sorry about I didn't close attention. Thank you.

**Comment by Ambro19** [comment]
Hi @vaithak, I just added some changes in manipulation.py and test_manipulation.py.
I wrote this testing scripts, but I'm having some errors which I will look at them later today. Thanks!
image

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @Ambro19,
I'll be taking over the PR review process from now on. Sorry for the inconvenience about that 😅.

I've skimmed the PR quickly and it seems like something weird happened with your PR, most likely you ran `pre-commit run --all-files` while linting. So there are lots of changes that are not related to your PR present. Would you mind reverting those changes?

**Comment by Ambro19** [comment]
Hey @KareemMAX, It's okay :) I am sorry for the delay I am just seeing your message right now. And sorry these unrelated commits. Let's me fix and get back to you. Thanks!

**Comment by Ambro19** [comment]
Hey @KareemMAX , I just reverted all the last three unrelated changes. Hope this fixes it. Please let me know. Thank you!

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @Ambro19,
Seems like the changes are still there sadly. I have 2 things in mind:
- Option 1: If you are good with git, you may reset and modify your branch and force push it so the unrelated changes are removed entirely.
- Option 2: Make a clean PR from main, that would involve some copying and pasting.

Hope those options would help, the latter would assign you to another engineer though. Feel free to mention me in a comment and I'll pick it right away

**Comment by Ambro19** [comment]
Hey @KareemMAX, thank you for your comment. I will do option 2.
